B2C Implementation Guide (#7310)
auth0/docs: articles/architecture-scenarios/implementation/b2c/authentication.md

---

> title: User Authentication

Reviewer: ```suggestion
title: Authentication
```

description: Understand how authentication works in your CIAM implementation.
toc: true
topics:
    - b2c
    - ciam
    - authentication
    - universal-login
contentType: concept
useCase:
    - authentication
---

> # User Authentication

Reviewer: ```suggestion
# Authentication
```

> In order to provide services to your users, you must be able to identify who those users are. This process is called authentication. You must provide your users with a way to verify that they are who they are, and there are a number of ways to do this; via social media accounts, username/password, and passwordless authentication to name but a few. In many cases it is also recommended that you go beyond a first factor for authenticating the user add a second factor as well.

Reviewer: ```suggestion
In order to provide services to your users, you must be able to identify who those users are. This process is called authentication. You must provide your users with a way to verify that they are who they are, and there are a number of ways to do this: via social media accounts, username/password, and passwordless authentication to name but a few. In many cases it is also recommended that you go beyond a first factor for authenticating the user add a second factor as well.
```

## Design considerations

> When designing the authentication experience, you will determine:

Reviewer: ```suggestion
When designing the authentication experience, you will need to consider:
```

* Where users will enter their credentials
* How you will keep user credentials safe
* How you will maintain your authentication system
* How you will provide password authentication for your users
* If you will integrate social logins for my users
* How you will prevent hackers from trying to log in as your users
* How you want to implement authentication in different kinds of applications using Auth0
* If you want to make login easy for your users when they come from different language backgrounds
* How you will provide a good user experience as you migrate away from my legacy authentication system

> The answers to these questions, and others, are detailed below in sections on [Universal Login](#universal-login), [user authentication](#user-authentication), [anomaly detection](#anomaly-detection), and [application integration](#application-integration).

Reviewer: This leaves it to the reader to have to figure out where the answers actually are. In line with our more prescriptive strategy we opted to explicitly guide the reader to where they need to go; hence the narrative from the original document (https://docs.google.com/document/d/1wHEmRsPVciJJVuyssBH3RyO1mLYIGqLj34-vhVFEODU/edit#bookmark=id.jw0r0xkd9gjo). Can we please restore the original narrative for this section. Thank you 👍

Author: Yes, I thought I did that. I'll go back and look at it again. It's really difficult to keep up if you keep changing the original source.

---
title: Authentication
description: Understand how authentication works in your CIAM implementation.
toc: true
topics:
    - b2c
    - ciam
    - authentication
    - universal-login
contentType: concept
useCase:
    - authentication
---
# Authentication

> In order to provide services to your users, you must be able to identify who those users are. This process is called authentication. You must provide your users with a way to verify that they are who they are, and there are a number of ways to do this: via social media accounts, username/password, and passwordless authentication to name but a few. In many cases it is also recommended that you go beyond a first factor for authenticating the user add a second factor as well.

Reviewer: ```suggestion
In order to provide services to your users, you must be able to identify who those users are. This process is called authentication. There are a number of ways to perform user authentication - via social media accounts, username/password, passwordless - and it's often recommended that you go beyond a first factor for authenticating the user by adding a second factor as well (a.k.a. Multi-factor Authentication).
```

> 

Reviewer: There was a "best practice" callout here that now seems to be missing (https://docs.google.com/document/d/1wHEmRsPVciJJVuyssBH3RyO1mLYIGqLj34-vhVFEODU/edit#bookmark=id.pi8vu9ftyaza).

Author: Added it back in.

When an anonymous user accesses your application, it can often be desirable for the application to discover if the user has already logged into a different application in the same family, or to remember this user even if the application is a SPA with no state. For example, if you can determine that the user is already logged in, you might decide to have the header of the application skip displaying a login button and instead have an account or profile menu for the user. To accomplish this you will want to utilize "silent authentication". Silent authentication will allow you to check to see if the user is logged in without prompting them to log in if they are not. Then the application can present a login button if necessary. If the user is logged in already, however, then you will receive tokens and will not have to present the user with a login button again.

::: warning
Checking for a login session by redirecting to Auth0 can be really helpful for your application, but if this will result in a lot of requests it is important to employ some sort of throttling mechanism to avoid latency and/or rate limiting.

> <%= include('../../_includes/_rate-limit-policy.md') %>

Reviewer: Might look better if we can include this as a continuation of the previous paragraph and not as a new paragraph?

Author: Fixed.

* Is your application a native application (mobile or desktop) and/or do you need a refresh token? See [Authorization Code Grant with PKCE](/api-auth/tutorials/authorization-code-grant-pkce) for more information. 

::: warning
Before you go live, you should ensure that only the grants that you are using for each application are enabled in your [configuration for the client](/applications/guides/update-grant-types-dashboard).
:::

### Implicit grant

If all your application needs is the ID Token and the application is browser-based, then you can always use the [implicit grant](/api-auth/tutorials/implicit-grant) to get your ID Token. This is a simple authentication flow and should be supported by your SDK (depending on the language you are developing in).

::: warning
If you need a [Refresh Token](/tokens/refresh-token/current) so that you can obtain a new Access Token or ID Token without having to re-authenticate the user, then you must use the [authorization code grant](/api-auth/tutorials/authorization-code-grant).
:::

### Authorization code grant (with or without PKCE)

> If your SDK only supports the authorization code grant, or you need an Access Token or Refresh Token, authorization code grant (with or without [PKCE](/flows/concepts/mobile-login-flow)) can also be used to retrieve an ID Token.  The authorization code grant includes an additional API call to exchange the code for a token which can result in additional unnecessary latency if all you need is the ID Token. In many cases the [hybrid flow](/api-auth/tutorials/hybrid-flow) is implemented to provide optimum access to the ID Token while still leveraging the authorization code grant workflow for the secure and safe retrieval of Access and Refresh Tokens.

Reviewer: ```suggestion
If your SDK only supports the authorization code grant, or you need an Access Token or Refresh Token, then authorization code grant (with or without [PKCE](/flows/concepts/mobile-login-flow)) can also be used to retrieve an ID Token.  The authorization code grant includes an additional API call to exchange the code for a token which can result in additional unnecessary latency if all you need is the ID Token. In many cases the [hybrid flow](/api-auth/tutorials/hybrid-flow) is implemented to provide optimum access to the ID Token while still leveraging the authorization code grant workflow for the secure and safe retrieval of Access and Refresh Tokens.
```
